src/main/java/com/direwolf20/buildinggadgets/common/util/blocks/RegionSnapshot.java

```diff
@@ -0,0 +1,361 @@
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.nbt.NBTUtil;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import net.minecraft.world.dimension.DimensionType;
+import net.minecraftforge.common.util.BlockSnapshot;
+import net.minecraftforge.common.util.Constants;
+import net.minecraftforge.common.util.TriPredicate;
+import net.minecraftforge.fml.server.ServerLifecycleHooks;
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.util.*;
+import java.util.function.BiPredicate;
+
+public class RegionSnapshot {
+
+    private static final String DIMENSION = "dim";
+    private static final String BLOCK_FRAMES = "frames";
+    private static final String BLOCK_PALETTES = "palettes";
+    private static final String TILE_DATA = "block_snapshots";
+    private static final String TILE_POS = "block_pos";
+    private static final String TILE_NBT = "block_nbt";
+
+    private static int nextPaletteeID(Object2IntMap<IBlockState> mapPalettes) {
+        // ID 0 is preoccupied for "nothing should be placed here", which is represented as Optional.empty()
+        return mapPalettes.size() + 1;
+    }
+
+    private static Optional<IBlockState> getPalettee(List<IBlockState> palettes, int id) {
+        return id == 0 ? Optional.empty() : Optional.of(palettes.get(id - 1));
+    }
+
+    private static NBTTagCompound serialize(NBTTagCompound tag, RegionSnapshot snapshot) {
+        tag.setInt(DIMENSION, snapshot.world.getDimension().getType().getId());
+        snapshot.region.serializeTo(tag);
+
+        // Palette serialization begin
+        NBTTagList palettes = new NBTTagList();
+        Object2IntMap<IBlockState> mapPalettes = new Object2IntOpenHashMap<>();
+        Set<IBlockState> recordedPalettes = new HashSet<>();
+
+        // "nothing should be here" block state
+        palettes.add(new NBTTagCompound());
+        // Palette serialization end
+
+        // Streaking: a way to record a number of repeating block states, using only 2 integers
+        // Storage frame: one or two numbers indicating a sequence of block states
+        // Regular frame: a type of storage frame; 1 number, representing the ID of the block state recorded
+        // Streak frame: a type of storage frame; 2 numbers, where the first one is positive, representing the ID of the repeating block state; the second one is negative, representing the number of repeating block states
+
+        // Complete storage frames, will be serialized into a NBTTagList later
+        IntList frames = new IntArrayList();
+
+        // The repeating block state, if any
+        Optional<IBlockState> lastStreak = Optional.empty();
+        // Number of repeating block states
+        int streak = 0;
```

Comment:
1. This can be fixed easily by replacing `<` with `<=`: (`for (int j = 0; j <= streaks; j++)`).
2. I think the original is semantically more appropriate, because there is no repeating block state before we even start. Additionally, leaving it at 0 can make `firstIteration` unnecessary, which is in our favor.

Reply:
If you do this, could you add a comment at the loop then, that there's only one situation, where it needs to run for a 0 streak?

Reply:
Need to correct myself: I think that's not going to work.

In the situation where you streak starts at one let's assume you've got a streak of 1+c with c a natrual number. This implies that your creation loop performed the increment c-times, aka there were streak=1+c blocks. If your deserialise loop now starts at 0 and runs while you're at <=c+1. This means, it will run c+2 times, aka streak+1. This is one more time then it should. (qed) 

This implies you need to do it consostenly as I alreadybattempted to convey. Either you always start at 1, and run your loop while <streak, or you always start at 0 and run your loop while <=. Mixing the 2 will not work (Proofing other combinations is analoge to above).

Reply:
This also implies you cannot replace first Iteration with it.

Reply:
From the standpoint of saving space it would make more sense to always start at 0 and use <= as this way you've got one more values of the streak byte used, allowing for it to have a by one longer streak then when starting at 1.
(please add comments that streaks then isn't the amount of blocks in a row, but one minus that...)

Reply:
Given the region contains sets of repeating block states. For every set `S`, our pointer will start one before `S`, and the counter `c` will get increased `#S` times. This means we will always end up with `c=#S` at one after `S`.
Based on your suggestion, I would store `streaks - 1` to NBT and read `streaks + 1` from NBT.

---

#### Notes
The "reset counter to 1" situation is yet another left-behind of multi-element-frame. I really got tripped when I saw it (my first reply), it is completely wrong
1. `for(int i = 0; i < n; i++)` will execute the statement body `n` times. Same for `for(int i = 1; i <= n; i++)` but not `for(int i = 0; i <= n; i++)` -- the latter will execute it `n+1` times.